core/src/main/scala/kafka/server/DelayedProduce.scala

```diff
@@ -124,6 +124,19 @@ class DelayedProduce(delayMs: Long,
   }
 }
 
+class SafeDelayedProduce(delayMs: Long,
+                         syncObject: Object,
+                         produceMetadata: ProduceMetadata,
+                         replicaManager: ReplicaManager,
+                         responseCallback: Map[TopicPartition, PartitionResponse] => Unit)
+  extends DelayedProduce(delayMs, produceMetadata, replicaManager, responseCallback) {
```

Comment:
Could we add a comment on when SafeDelayedProduce is needed (e.g. avoid deadlock)?